### ISSUE TITLE
[FEATURE] "Show Items in Sub-Folders" and asset search streamline

### ISSUE THREAD

**Issue by RiederAlex** [ezEngine/ezEngine#1315]
Currently if the editor is started, "Show Items in Sub-Folders" is always active in the asset browser.
![subfolders](https://github.com/ezEngine/ezEngine/assets/14026803/619117c5-2622-4b32-9bdb-6a67cc05118e)

I think this is a bit confusing, as this is not how most file managers/Unity/Unreal work by default - which I assume is what a user would expect. So I think it would improve the onboarding experience if the default was changed to false there.

If this flag is now disabled, searching becomes a bit more cumbersome - but here also copying the behavior from file managers/Unity/Unreal could lead to an improvement, by temporarily acting as if the checkbox is enabled as soon as a user types something in the searchbar.

This new behavior would look like this - notice how "Show Items in Sub-Folders" is disabled, but it still shows search results of all subfolders.

https://github.com/ezEngine/ezEngine/assets/14026803/1aaae5ab-68da-4ad1-b6a1-560780d7d1e8

So the feature request would be:

- Disable this checkbox per default
- Optionally implement saving the state of it across editor runs, now it resets everytime the editor is restarted
- Implement new behavior for searching shown in the video

If this is accepted, I would take over the implementation of this.

**Comment by jankrassnigg** [comment]
I disagree with this. Unity's asset browser experience is, in my opinion, pretty awful and only because people are used to it, I'm not willing to make our asset browser worse.

It is, in fact, a deliberate decision NOT to save this state across runs, because then people coming from Unity would immediately switch it, because they are used to that, and then complain that their experience is degraded.

All the options in the ez asset browser are filters that combine together. One is a type filter, one is a path filter and one is a custom text expression filter. The additional options in the context menu break this behavior a bit, but in rare cases they are useful to have. But usually they are not needed. I need to switch the "Show items in sub-folders" option maybe once a month.
The text filter is not a "search" and it is not supposed to completely change the behavior. The concepts are all very simple and once you figured out how they work, it is extremely powerful.

I have observed people coming from Unity working with a similar implementation in another engine, and they are used to expanding folders all the time. Need to open that mesh ... oh it was somewhere in this folder, let me go through the hierarchy, 20 folders deep ... ah here it is.... 
People literally spend a minute just going through folders, because they are used to doing it this way. Sure, they can also "search" for things, but often you work with assets that belong together and thus have different names, so going to the actual folder is preferable.
In ez you should get used to navigate differently. Filter by names for a thing that you know. If you then want to restrict the result to all the files in the same folder, right click on and asset and select "Filter path to this". Viola, two actions, you are done. Since assets in sub-folders are shown, you can organize your files much more fine grained, but you won't have problems finding them from the asset browser, no hunting down in which folder exactly they are, which is such a huge waste of time in Unity.

If you automatically switch this behavior on and off when you type something into the text filter, this is a complete mode switch, which is very irritating for users. Also you then have to deal with the fact that people may NOT want this in some situations, now you need more options to auto switch or not, under some circumstances. It all becomes even more confusing or less useful.

Now, there are a few details that could be improved. For example the option "Show items in sub-folders" isn't easy to find (same for the other option from the context menu). We could make them more accessible, for example by adding them as toolbar buttons in the asset browser, so that people can see that this option exists and switch back and forth quicker.

Sorry for this rant, but this isn't the first time I've had this discussion and it always boils down to "but I'm used to this from Unity". There is feedback about making something more useful, and there is feedback about making something more similar to something else. And unfortunately Unity is often a prime example for how not to do things (don't get me started on their god-awful way of using their Inspector for everything), but people are so used to this, that they don't want to get used to something else, even though it would actually make their workflow better, if they adjusted a little bit and actually utilized the strengths of other tools.

One other example for this is, that Unity always shows the asset browser on the bottom, which is actually not that convenient with wide-screen monitors. If you dock it on the right, or even just undock it and move it to a second screen, you get much more screen real-estate to work with.

**Comment by RiederAlex** [comment]
Let me preface my answer with the fact that this is your project, I really like working on/with it and my intent is to improve ez, not to tell you guys "how it should be done".
Additionally let me also state that I as well question many UI/UX decisions of Unity/Unreal and that I, as well, have experience working on larger projects.

### Convention/Expectation
From my perspective/experience with non-technical users expect what I described - and I'm not just talking about recovering Unity devs here - I'm also talking about how file browsers work in any operating system.
And from a user's perspective the AssetBrowser is basically a file browser - I have files, I have folders, I have a search bar = File browser.

Even with knowing how it works, its often confusing and counter intuitive - e.g. I drag & drop an asset into a folder - and it does something - but the picture of the asset is still at the same place? Did I move it now? Or did the Drag&Drop fail for some reason?

### Scaling

Expanding folders by default removes one of the use-cases for having folders in the first place - I want folders to act as filters/visibility barries, to NOT see everything all at once. 

This means the larger the project gets, the less usefull will be the non-filtered view of top-level folders, as it shows so many assets a user is most likely not interested in.

This continues with sub-folders, as they also grow, making this view less and less relevant, because it shows so much I dont want at this point in time.

So I have to use the search to drill down to what I need, and right click -> Filter to this path, which is what usually everybody does on bigger projects. And if the filter is cleared then, and the current filtered path has a lots of sub-folders with different assets, I again see too much what is not relevant for me - Didn't I just select this folder and filtered by it? Why am I still seeing so much assets? Are they in my current folder or in a subfolder?

---

To answer some of your points more directly:

> The text filter is not a "search" and it is not supposed to completely change the behavior.

As state above, as a non-technical user it sure feels that way - it essentially is a search.

> let me go through the hierarchy, 20 folders deep ... ah here it is.... [...] so going to the actual folder is preferable.

This is a Unity specific issue, and ez already has the exact same solution as Unreal - which is RightClick -> Filter to this path.
So the flow would be - I search/filter for the asset I know -> RightClick, Filter to this path and I'm at exactly the point where I want to be.
This is no different with my proposal than with the current solution - both are two actions, both in the selected folder & showing me only assets I actually want to see.

> If you automatically switch this behavior on and off when you type something into the text filter, ...

As state above, as a non-technical user I actually expect it to work this way, as for such users its a file browser.

---

To summarize, I'm not arguing against the current setup (Filter by Asset Type, Select a folder on the left side & search for assets), and the proposed changes do not break this behavior in any way, its really just about auto expanding all folder contents which breaks user expectations & making the Asset Browser behave more like a traditional file browser.

**Comment by jankrassnigg** [comment]
Ez does use the convention that assets in folders ending with "_data" are always hidden, unless you navigate directly into the folder or switch the flag to show them anyway. This is because there are many assets that belong to exactly once other asset and are rarely needed directly. Curious to hear your opinion about that feature, because with your proposed change this would become either irrelevant or more complicated.

The drag drop stuff is relatively new and you are right that it doesn't work that well with this approach. The asset browser has become more and more a file browser, which wasn't the original intent.

If you have a branch with a working change, I'd be willing to give it a try. I'd not be against making it an experimental feature that can be enabled in the preferences, so that we can test a bit longer whether this is better or not. I am very used to how it is right now and think it is better, but you do raise a few good points. 

And as I said, it would also be useful to make the show option more accessible in general. If there was a checkable button on the toolbar, people could switch it easily, and if you switch the mode automatically on text search, the button could get disabled and checked, to communicate to the user that this mode is now always active. One could also add a tooltip with some explanation.

**Comment by RiederAlex** [comment]
I've implemented a draft version in a branch - including adding a toggle button (which defaults to off). Search now always looks into sub-folders and disables the button temporarily.

It for sure needs a bit more work, like proper icons, but it should be enough to play around with it:
https://github.com/RiederAlex/ezEngine/tree/user/ar/assetbrowserchange

Regarding the "_data" folder issue - I think this can potentially stay as-is, as from a user-standpoint I won't be interested in them most of the time?
There still exists the the "Show Items in hidden folder" setting (which in my branch is a bit cumbersome to activate right now), which expert users can enable to see even more information.

**Comment by jankrassnigg** [comment]
Great, thank you! I'll take a look over the weekend or early next week, I'm currently on the road back from vacation.

**Comment by jankrassnigg** [comment]
Just some quick feedback, haven't had much time yet.

I'd add this:

![image](https://github.com/ezEngine/ezEngine/assets/6001174/33178ca9-7a88-4f3d-a2bf-844782989d2a)

That gives visual feedback that this mode is now force enabled when text is entered.

Also when the root level is activated we still show all files, I guess this is a pre-existing bug:

![image](https://github.com/ezEngine/ezEngine/assets/6001174/a3a4443c-fd83-4127-adef-f19c57cdc57e)

**Comment by jankrassnigg** [comment]
I think the proposed change is decent. Still need to get used to it, but generally I think we can make this change, so feel free to open a PR.
